doc/md/fundamentals/3-declarations/7-module-declarations.md

```diff
@@ -0,0 +1,111 @@
+---
+sidebar_position: 7
+---
+
+# Module declarations
+
+In Motoko, a **module** is a collection of related types, values, and logic grouped under a single namespace. Unlike actors and classes, modules cannot hold or mutate state, making them ideal for reusing side-effect-free code.
+
+Modules are mainly used to build libraries, such as those in the [base library](https://internetcomputer.org/docs/motoko/main/base/) or packages available through [Mops, the Motoko package manager](https://mops.one).
+
+A module in Motoko can define:
+
+- Public types.
+- Public functions, both synchronous and asynchronous.
+- Private types and internal logic that is not exposed outside the namespace.
+- Nested modules.
+
+Modules provide encapsulation and structure, allowing code to be organized and reused across different parts of an application.
+
+```motoko no-repl
+import Array "mo:base/Array";
+
+module Matrix {
+  // Define what a matrix is - a 2D array of Nat
+  public type Matrix = [[Nat]];
+
+  // Create a matrix filled with a default value
+  public func init(rows : Nat, cols : Nat, default : Nat) : Matrix {
+    Array.tabulate<[Nat]>(rows, func (i : Nat) {
+      Array.tabulate<Nat>(cols, func (j : Nat) {
+        default
+      })
+    })
+  };
+
+  // Create an identity matrix (ones on diagonal, zeros elsewhere)
+  public func identity(n : Nat) : Matrix {
+    Array.tabulate<[Nat]>(n, func (i : Nat) {
+      Array.tabulate<Nat>(n, func (j : Nat) {
+        if (i == j) { 1 } else { 0 }
+      })
+    })
+  };
+
+  // Transpose a matrix (flip rows and columns)
+  public func transpose(m : Matrix) : Matrix {
+    if (m.size() == 0) return [];
+    let rows = m.size();
+    let cols = m[0].size();
+    Array.tabulate<[Nat]>(cols, func (j: Nat) {
+      Array.tabulate<Nat>(rows, func (i:Nat) {
+        m[i][j]
+      })
+    })
+  };
+};
+```
+
+This module declares:
+
+- A public type `Matrix.Matrix` for representing a 2D array of natural numbers `[[Nat]]`.
+- Functions `Matrix.init()`, `Matrix.identity()`, and `Matrix.transpose()` for matrix operations.
+
+:::note Functional programming style
+This module follows a functional programming approach: all operations return new matrices instead of modifying existing ones. This immutable design improves predictability and minimizes side effects.
```

Comment:
```suggestion
This module follows a functional programming approach: all operations return new matrices instead of modifying existing ones. The lack of side-effects makes this style of programming easier to reason about.
```